test/extensions/quic_listeners/quiche/active_quic_listener_test.cc

```diff
@@ -84,102 +184,81 @@ class ActiveQuicListenerTest : public testing::TestWithParam<Network::Address::I
-      new quic::QuicSignedServerConfig);
-  quic::QuicCompressedCertsCache cache(
-      quic::QuicCompressedCertsCache::kQuicCompressedCertsCacheSize);
-  quic::test::crypto_test_utils::GenerateFullCHLO(
-      chlo, &ActiveQuicListenerPeer::crypto_config(*quic_listener_),
-      envoyAddressInstanceToQuicSocketAddress(listen_socket_->localAddress()),
-      envoyAddressInstanceToQuicSocketAddress(client_socket_->localAddress()),
-      quic::AllSupportedVersions()[0].transport_version, &clock, signed_config, &cache, &full_chlo);
-  // Overwrite version label to highest current supported version.
-  full_chlo.SetVersion(quic::kVER, quic::CurrentSupportedVersions()[0]);
-  quic::QuicConfig quic_config;
-  quic_config.ToHandshakeMessage(&full_chlo, quic::CurrentSupportedVersions()[0].transport_version);
-
-  std::string packet_content(full_chlo.GetSerialized().AsStringPiece());
-  auto encrypted_packet =
-      std::unique_ptr<quic::QuicEncryptedPacket>(quic::test::ConstructEncryptedPacket(
-          connection_id, quic::EmptyQuicConnectionId(), /*version_flag=*/true, /*reset_flag*/ false,
-          /*packet_number=*/1, packet_content));
-
-  Buffer::RawSlice first_slice{reinterpret_cast<void*>(const_cast<char*>(encrypted_packet->data())),
-                               encrypted_packet->length()};
-  // Send a full CHLO to finish 0-RTT handshake.
-  auto send_rc = Network::Utility::writeToSocket(client_socket_->ioHandle(), &first_slice, 1,
-                                                 nullptr, *listen_socket_->localAddress());
-  ASSERT_EQ(encrypted_packet->length(), send_rc.rc_);
-
-  EXPECT_CALL(listener_config_, filterChainManager());
-  EXPECT_CALL(filter_chain_manager_, findFilterChain(_));
-  EXPECT_CALL(filter_chain_, networkFilterFactories());
-  EXPECT_CALL(listener_config_, filterChainFactory());
-  EXPECT_CALL(listener_config_.filter_chain_factory_, createNetworkFilterChain(_, _));
-  EXPECT_CALL(*read_filter_, onNewConnection())
-      // Stop iteration to avoid calling getRead/WriteBuffer().
-      .WillOnce(Invoke([this]() {
-        dispatcher_->exit();
-        return Network::FilterStatus::StopIteration;
-      }));
-  EXPECT_CALL(network_connection_callbacks_, onEvent(Network::ConnectionEvent::Connected));
-
-  dispatcher_->run(Event::Dispatcher::RunType::Block);
-
-  Buffer::InstancePtr result_buffer(new Buffer::OwnedImpl());
-  const uint64_t bytes_to_read = 11;
-  uint64_t bytes_read = 0;
-  int retry = 0;
-
-  do {
-    Api::IoCallUint64Result result =
-        result_buffer->read(client_socket_->ioHandle(), bytes_to_read - bytes_read);
-
-    if (result.ok()) {
-      bytes_read += result.rc_;
-    } else if (retry == 10 || result.err_->getErrorCode() != Api::IoError::IoErrorCode::Again) {
-      break;
-    }
+  ConfigureMocks(/* connection_count = */ 1);
+  GenerateCHLO(quic::test::TestConnectionId(1));
+  dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
```

Comment:
Would NonBlock here cause data race where data arrives after we exit the loop? Can we still use Block and exit() instead? Actually this test maybe by nature flaky because we can't guarantee that the 16 CHLOs arrive in one event loop. Can you test with --runs_per_test to see if it flakes?

Reply:
Sorry, I do not understand the event loop well enough to be able to tell if data could possibly arrive after we exit the loop.  If you think it is possible, I could use Block and exit().

I also cannot say if this test could be flaky.  If you think it is possible, then we could lower the buffered CHLO limit to 1 and feed 2, that would be an improvement because 2 must have a better chance to arrive in the same event loop than 16.

Reply:
Can you try --runs_per_test to see if it's flaky first? My understanding about NonBlock is to run one loop and exit. If there is no active IO events, that's it. 

Lowering the limit sounds reasonable. But again let's make sure the test doesn't flake before check in.

Reply:
Actually, I am not able to run this test locally.  It crashes with segmentation fault.  Both direct bazel and when using a docker container.  Does the same on HEAD.

Reply:
NonBlock guarantees that the event loop will only be run once.  Otherwise all CHLOs would be processed, because the limit kNumSessionsToCreatePerEpoll is per event loop.  Also Block would require calling exit() from the last read_filter.onNewConnection() mock call but not from earlier ones, which is certainly doable but seems to add complexity unnecessarily to test code.

Reply:
Agree that Block and exit() doesn't satisfy our need. Are we sure that there is no race between receiving the packet and dispatcher.run() now?